ctapipe/instrument/camera.py

```diff
@@ -308,32 +321,74 @@ def __str__(self):
 
     @lazyproperty
     def neighbors(self):
-        """" only calculate neighbors when needed or if not already
-        calculated"""
-
-        # return pre-calculated ones (e.g. those that were passed in during
-        # the object construction) if they exist
-        if self._precalculated_neighbors is not None:
-            return self._precalculated_neighbors
-
-        # otherwise compute the neighbors from the pixel list
-        dist = _get_min_pixel_seperation(self.pix_x, self.pix_y)
-
-        neighbors = _find_neighbor_pixels(
-            self.pix_x.value,
-            self.pix_y.value,
-            rad=1.4 * dist.value
-        )
-
-        return neighbors
+        '''A list of the neighbors pixel_ids for each pixel'''
+        return [np.where(r)[0].tolist() for r in self.neighbor_matrix]
 
     @lazyproperty
     def neighbor_matrix(self):
-        return _neighbor_list_to_matrix(self.neighbors)
+        return self.neighbor_matrix_sparse.A
 
     @lazyproperty
     def neighbor_matrix_sparse(self):
-        return csr_matrix(self.neighbor_matrix)
+        if self._neighbors is not None:
+            return self._neighbors
+        else:
+            return self.calc_pixel_neighbors(diagonal=False)
+
+    def calc_pixel_neighbors(self, diagonal=False):
+        '''
+        Calculate the neighbors of pixels using
+        a kdtree for nearest neighbor lookup.
+
+        Parameters
+        ----------
+        diagonal: bool
+            If rectangular geometry, also add diagonal neighbors
+        '''
+        neighbors = lil_matrix((self.n_pixels, self.n_pixels), dtype=bool)
+
+        if self.pix_type.startswith('hex'):
+            k = 7
+            radius = 1.4
+            norm = 2  # use L2 norm for hex
+        else:
+
+            # if diagonal should count as neighbor, we
+            # need to find at most 8 neighbors with a max L2 distance
+            # < than 2 * the pixel size, else 4 neigbors with max L1 distance
+            # < 2 pixel size
+
+            if diagonal:
+                k = 9
+                norm = 2
+                radius = 1.95
+            else:
+                k = 5
+                radius = 1.5
+                norm = 1
+
+        for i, pixel in enumerate(self._kdtree.data):
+            d, n = self._kdtree.query(pixel, k=k, p=norm)
+
+            # remove self-reference
+            d = d[1:]
+            n = n[1:]
+
+            # remove too far away pixels
+            mask = d < radius * np.min(d)
+            neighbors[i, n[mask]] = True
+
+        # filter annoying deprecation warning from within scipy
+        # scipy still uses np.matrix in scipy.sparse, but we do not
+        # explicitly use any feature of np.matrix, so we can ignore this here
+        with warnings.catch_warnings():
+            warnings.filterwarnings('ignore', category=PendingDeprecationWarning)
+            if (neighbors.T != neighbors).sum() > 0:
```

Comment:
Isn't this `.sum() > 0` actually an `.any()` ? In english this would then read roughly like: 

if the transposed neighbors is not equal to neighbors, anywhere, then...

Reply:
Sparse matrices do not have any, unfortunately.

Reply:
ok  ... sorry .. should have checked that myself.